src/posts/abstracting-the-frontend/index.md

```diff
@@ -0,0 +1,124 @@
+across all the types of gadgets and gizmos that exist out there.
+
+_**Quick disclaimer:** Perhaps these concepts can be sorted into something more pragmatic than layers,
+the important part is the separation of concerns._
+
+![Front End Layers](./diagram.png)
+_An illustration of the Front End Layers_
+
+#### Design system
+
+As front-end developers, UI and UX designers we are responsible for the visual representation and interaction of our products.
+What we present to our users should always be the number one concern of any front-end unit. Building a design system early
+and sticking to it as if it was holy law is imperative to succeed in scaling your products.
+The implications of what a design system is may vary but these are the assertions that I believe we should make about them.
+
+1. The core of a design system should be reusable across products. A company's products don't have to look the same,
+   but they should feel the same.
+2. Component libraries should exist that strictly follow the rules of the design system.
+3. A component library should not consume remote data directly.
+4. A design system's documentation should be available to the public.
+
+**Note:** Point 3 makes the layer approach a bit awkward, but I hope we can agree on the general philosophy.
+
+#### Data aggregate
+
+If we ever assumed that the backend team is only concerned delivering data to our front ends, this is no longer the case.
+Backend developers building API's have their own group of stakeholders to worry about, integrators. This means that there will be
+certain compromises in API design. Thus, having a middleware where we can create new datasets tailored
+to the front end can prove quite useful.
+
+A front end that does not communicate with a data source is a rare sight these days and I believe that the number of data points a
+single client side application will have to interact with will only grow. A single resource in our front end might consist of
+multiple different sources of data, so in order to reduce complexity in our application code, we will want to host an aggregate of the
+data our front ends interact with so that we can simplify what our client will consume.
+
+#### The application
+
+This one might not come as a big surprise. This is the layer that separates a fridge from a smart watch.
+But provided that we have the above layers, maintaining our applications should now be a less daunting task.
+I had thought about splitting this up into a logic and view layer, but being a firm believer in colocation I feel that
+sharing logic between applications is not something that needs to be asserted in order to succeed.
+There are people out there far smarter than me that will deep dive into what makes a scalable front-end application.
+
+Check out Kent C. Dodds' post on [Colocation](https://kentcdodds.com/blog/colocation) if you're interested in learning more about the principle.
```

Comment:
And of course I disagree completely with his approach to comments, but that could be a real-life discussion if youre up for it.

Reply:
I also disagree with Kent's approach to comments. I wish he would have picked another analogy to explain the idea of colocation. Perhaps I will just point to a more formal definition.